### PR TITLE
Partly fixes #2062: part5b refactor register provider

### LINKED ISSUE #2062: Refactor pywbem mock to separate components
We propose to refactor the mocker to:

a. Separate it into the components that would conceptually represent a WBEM server including:

1. **repository store class.** This component will manage the data store itself including adding, modifying, deleting CIM classes, instance, qualifier declarations and managing the creation/deletion/enumeration of the namespaces represented in the  data store.  The objective of this component is to:
a. Physically store the objects (CIM classes, instances,qualifier declarations and namespace information)
b. Provide APIs so that the server repository class and providers can add items to the store, and retrieve items from the store.  
c. Separate the store implementation itself (ex. current dictionaries or possible databases in the future from the logical repository.
d. Provide validation of the objects stored to assure the data in the store is valid.

This API is effectively what exists in the mof_compiler BaseRepositoryConnection class today with the addition of the namespace management methods.  This replaces the current _mockmofwebemconnection.py module in pywbem_mock.

 2. **server repository class** - The set of APIs that represent  the calls to the what is conceptually a WBEM server repository. This is all of the methods in cim_operations except for InvokeMethod, pull operations, (and possibly query) which we do not implement within  a repository. It also adds the methods for managing namespaces. This class will manipulate objects in the datastore class using the APIs in that class and manage the testing/manipulation of incoming cim operations to commucate between the objects in the data store and the user for repository objects.  This will include most of the methods currently contained in the mock calls in _wbemconnection_mock.py
    
3. **wbem server class**.  This class represents the server which separates requests into providers and the repository.  It include all calls defined in cim_operations.py. Without any providers, it simply passes operations on to the repository class.  TODO sort out issues of the InvokeMethod implementation which should be part of the providers whereas today they are sort of part of the repository.  This is the class that will be surfaced with the definitons that are today as FakedWBEMConnection and is the class used by the user to define the wbem connection and to insert data into the repositoy the repository.
    
4. **MOCKER**. The mocker which mocks the imethod_call and calls the appropriate server methods  This class would include the mock code and the implementation of each of what is now the fake_... methods that simply call their WBEM server class counterparts. For example _fake_enumerateclasses would simply call the WBEMServer.EnumerateClasses with the parameters defined by cim_operations.py API. It will also map the responses into the form required by imethod_call so that the wbem_server class returns values the same as the cim_operations.py methods.

5. **Provider class.** This class that would provide the API for implementing providers.  This will provide the template for any future provider implementations.  This is a base class from which individual providers could be built.

### PR DESCRIPTION
**Pull request by KSchopmeyer**
See commit message for details.

Marked WIP for the moment but it passes tests.

**DISCUSSION:** In this code I ignore  attribute error in call to post_register_setup(conn).  Not sure whether we should ignore (they supply if they want to) or error(they must supply at least a dummy). Since this is server side code, I figured that they will figure it out.

### CODE REVIEW

**Comment by andy-maier** [comment]
Rebase on master

**Comment by andy-maier** [comment]
Please rebase on latest master, deleting the commits that belong to the other PR, and the conflicts will go away.

File: pywbem_mock/_wbemconnection_mock.py [hunk 1]
```diff
@@ -307,6 +311,19 @@ class must contain variables `provider_type` and
...
+                        "type: {1!A} namespaces:{2!A}",
+                        provider.__class__.__name__,
+                        ", ".join(provider_classnames),
+                        provider_type, ", ".join(namespaces))
+
+            try:
+                provider.post_register_setup(conn)
+            except AttributeError:
+                if verbose:
+                    _format("Register_Provider post_register_setup not found")
```
Comment (andy-maier) [inline pywbem_mock/_wbemconnection_mock.py:325]: I don't think we should require that a provider has this method. It should be optional: If defined, it is called.

Comment (KSchopmeyer) [reply pywbem_mock/_wbemconnection_mock.py:325]: Good.  No warning, etc. if missing.

**Comment by KSchopmeyer** [comment]
I rebased on master and cleaned up from where I had multiple commits of my own.  Also rewrote the commit message.

Sorry but I cannot figure out how to  remove the two old commits.  Rebasing on master does not do it. Suggestions welcome.

Note that we simply try the post_register_setup and ignore if there is an attribute error.  Since this is a mocker, if there are other errors, it should probably fail or at least capture those exceptions so I do not unregister the provider upon error.

**Comment by andy-maier** [comment]
I squashed the 3 commits into one and will merge it when the tests complete.

**Pull request merged**